File: book/continuous-time/c#/Shared/Shapes.cs
using System.Windows;
using System.Windows.Media;
using Sodium.Frp;

namespace Shared
{
    public class Shapes
    {
        public static Behavior<DrawableDelegate> Circle(Color color)
        {
            return Behavior.Constant(new DrawableDelegate((d, h, o, s) =>
            {
                double radius = s;
                d.DrawEllipse(new SolidColorBrush(color), new Pen(Brushes.Black, 1), new Point(o.X, h - o.Y), radius, radius);
            }));
        }

        public static Behavior<DrawableDelegate> Scale(Behavior<DrawableDelegate> drawable, Behavior<double> scale)
        {
            return drawable.Lift(scale, (dr, ns) =>
                new DrawableDelegate((d, h, o, s) => dr(d, h, o, s * ns)));
        }

        public static Behavior<DrawableDelegate> Translate(Behavior<DrawableDelegate> drawable, Behavior<Point> offset)
        {
            return drawable.Lift(offset, (dr, no) =>
                new DrawableDelegate((d, h, o, s) => dr(d, h, no.Add(o.Multiply(s)), s)));
        }

        public static Behavior<DrawableDelegate> Over(Behavior<DrawableDelegate> a, Behavior<DrawableDelegate> b)
        {
            return a.Lift(b, (dra, drb) => new DrawableDelegate((d, h, o, s) =>
            {
                drb(d, h, o, s);
                dra(d, h, o, s);
            }));
        }
    }
}

File: book/continuous-time/f#/Shared/Shapes.fs
module Shared.Shapes

open System.Windows
open System.Windows.Media
open Sodium.Frp

let circle color = constantB (DrawableDelegate (fun d h o s ->
    let radius = s
    d.DrawEllipse(SolidColorBrush(color), Pen(Brushes.Black, 1.0), Point(o.X, h - o.Y), radius, radius)))

let scale drawable scale =
    let getScaleDrawableDelegate dr ns = DrawableDelegate (fun d h o s -> DrawableDelegate.invoke dr d h o (s * ns))
    lift2B getScaleDrawableDelegate (drawable, scale)

let translate drawable offset =
    let getTranslateDrawableDelegate dr no = DrawableDelegate (fun d h o s -> DrawableDelegate.invoke dr d h (Point.add no (Point.multiply o s)) s)
    lift2B getTranslateDrawableDelegate (drawable, offset)

let over a b =
    let getOverDrawableDelegate a b = DrawableDelegate (fun d h o s ->
        DrawableDelegate.invoke b d h o s
        DrawableDelegate.invoke a d h o s)
    lift2B getOverDrawableDelegate (a, b)